lighthouse-core/gather/driver.js

```diff
@@ -423,11 +423,15 @@ class Driver {
    * will be evaluated in a content script that has access to the page's DOM but whose JavaScript state
    * is completely separate.
    * Returns a promise that resolves on the expression's value.
-   * @param {string} expression
-   * @param {{useIsolation?: boolean}=} options
-   * @return {Promise<*>}
+   * @template T, R
+   * @param {string | ((...args: T[]) => R)} expression
+   * @param {{useIsolation?: boolean, mode?: 'iffe'|'function', args?: T[], deps?: (Function|string)[]}=} options
+   * @return {Promise<R>}
```

Comment:
Thanks for the great feedback!

> js-style "overloading" tends to grow in implementation complications over time and is harder to understand as a caller (e.g. options could be included with a string expression, but would it do anything in that case? Now we have to document that, etc)

I took a second pass to incorporate this feedback, and added some docs. Running to a meeting, will respond to the rest later.

Reply:
Would you prefer splitting this out to `evaluate(fn, ...)` and `evaluateString(code, ...)` (alias for `evalauteAsync`)?

Reply:
rather than add another `evaluate*()` hop with `_evaluate`, what do you think about keeping it called `evaluateAsync`, so

`evaluate(fn, ...)` calls `evaluateAsync(string, opts)` calls `_evaluateInContext(string, contextId)`

then in the next breaking change we either incorporate the guts of `evaluateAsync` into `evaluate` so it's back to only two functions, or we rename `evaluateAsync` to `evaluateString` or whatever if we do want to keep it around for folks to use?

I see the reason for doing it the `_evaluate` way, but the number of evaluate layers doesn't seem worth it when external to driver we still only have `evaluate()` and `evaluateAsync` either way.

Reply:
done

Reply:
I'm a little confused about the purpose of `_evaluate`. Is it for a possible patch where `evaluateAsync` is removed but we still need a function for executing short expressions?

Reply:
I think I just forgot to delete `_evaluate`.